common/src/main/java/io/netty/util/NettyRuntime.java

```
@@ -0,0 +1,74 @@
+
+    /**
+     * No public constructor to prevent instances from being created.
+     */
+    private NettyRuntime() {
+    }
+
+    /**
+     * Set the number of available processors.
+     *
+     * @param availableProcessors the number of available processors
+     * @throws IllegalArgumentException if the specified number of available processors is non-positive
+     * @throws IllegalStateException    if the number of available processors is already configured
+     */
+    public static void setAvailableProcessors(final int availableProcessors) {
+        if (availableProcessors <= 0) {
+            throw new IllegalArgumentException(
+                    "availableProcessors must be positive but was [" + availableProcessors + "]");
+        }
+        if (!NettyRuntime.availableProcessors.compareAndSet(0, availableProcessors)) {
+            throw new IllegalStateException(
+                    "availableProcessors is already set to [" + NettyRuntime.availableProcessors.get() + "], " +
+                            "rejecting [" + availableProcessors + "]");
+        }
+    }
+
+    /**
+     * Get the configured number of available processors. The default is {@link Runtime#availableProcessors()}. This
+     * can be overridden by setting the system property "io.netty.availableProcessors" or by invoking
+     * {@link #setAvailableProcessors(int)} before any calls to {@link #availableProcessors()}.
+     *
+     * @return the configured number of available processors
+     */
+    @SuppressForbidden(reason = "to obtain default number of available processors")
+    public static int availableProcessors() {
+        if (NettyRuntime.availableProcessors.get() == 0) {
+            final int availableProcessors =
+                    SystemPropertyUtil.getInt(
+                            "io.netty.availableProcessors",
+                            Runtime.getRuntime().availableProcessors());
+            setAvailableProcessors(availableProcessors);
```

I think that you're optimizing for the wrong thing. You're optimizing for a scenario that shouldn't happen, it should be considered a bug if it happens. By optimizing for that scenario, you're permitting bad racing scenarios to silently succeed (namely, a get racing with a set). My implementation prevents all races, it treats them as application bugs. I maintain that my implementation should be be preferred.